octodns/provider/azuredns.py

```diff
@@ -0,0 +1,369 @@
+                    string += 'Record: {}; '.format(self.params[char].__dict__)
+        return string
+
+
+def _validate_per(string):
+    return string if string.endswith('.') else string + '.'
+
+
+class AzureProvider(BaseProvider):
+    '''
+    Azure DNS Provider
+
+    azuredns.py:
+        class: octodns.provider.azuredns.AzureProvider
+        # Current support of authentication of access to Azure services only
+        # includes using a Service Principal:
+        # https://docs.microsoft.com/en-us/azure/azure-resource-manager/
+        #                        resource-group-create-service-principal-portal
+        # The Azure Active Directory Application ID (aka client ID) req:
+        client_id:
+        # Authentication Key Value req:
+        key:
+        # Directory ID (referred to tenant ID) req:
+        directory_id:
+        # Subscription ID req:
+        sub_id:
+        # Resource Group name req:
+        resource_group:
+
+        TODO: change config file to use env vars instead of hard-coded keys
+
+        personal notes: testing: test authentication vars located in
+                /home/t-hehwan/vars.txt
+    '''
+    SUPPORTS_GEO = False
+    SUPPORTS = set(('A', 'AAAA', 'CNAME', 'MX', 'NS', 'PTR', 'SRV', 'TXT'))
+
+    def __init__(self, id, client_id, key, directory_id, sub_id,
+                 resource_group, *args, **kwargs):
+        self.log = logging.getLogger('AzureProvider[{}]'.format(id))
+        self.log.debug('__init__: id=%s, client_id=%s, '
+                       'key=***, directory_id:%s', id, client_id, directory_id)
+        super(AzureProvider, self).__init__(id, *args, **kwargs)
+
+        credentials = ServicePrincipalCredentials(
+            client_id, secret=key, tenant=directory_id
+        )
+        self._dns_client = DnsManagementClient(credentials, sub_id)
+        self._resource_group = resource_group
+        self._azure_zones = set()
+
+    def _populate_zones(self):
+        self.log.debug('azure_zones: loading')
+        list_zones = self._dns_client.zones.list_by_resource_group
+        for zone in list_zones(self._resource_group):
+            self._azure_zones.add(zone.name)
+
+    def _check_zone(self, name, create=False):
+        '''
+            Checks whether a zone specified in a source exist in Azure server.
+            Note that Azure zones omit end '.' eg: contoso.com vs contoso.com.
+            Returns the name if it exists.
+
+            :param name: Name of a zone to checks
+            :type  name: str
+            :param create: If True, creates the zone of that name.
+            :type  create: bool
+
+            :type return: str or None
+        '''
+        self.log.debug('_check_zone: name=%s', name)
+        try:
+            if name in self._azure_zones:
+                return name
+            self._dns_client.zones.get(self._resource_group, name)
+            self._azure_zones.add(name)
+            return name
+        except:
```

Comment:
This should catch a specific exception type that has to do with the zone not existing rather than all possible exceptions. Hopefully the Azure client lib has clean/specific does not exist exception, but if not there's probably :crossed_fingers: a generic client exception that's throw where you can then look at the exception message.  

Catching any exception like this will hide all sort of problems and make tracking down issues harder for users. E.g. if the problem is an auth error it'll take both the get & create calls before raising. This would even catch problems in the code itself, `KeyError`, `ValueError`, ... anything, and try and still try and create the zone.

For a more specific problematic case an intermittent error (network, azure hiccup etc.) that fails the fetch in a call where `create=True` would then try and create a zone that probably already exists. That'd be confusing and potentially scary for a user.